WT-11126 Add configuration compilation to speed up WT_SESSION::begin_transaction

ddanderson opened the pull request:
This initial ticket creates an infrastructure to compile configuration strings. This can be used by applications to produce substitute strings that can used in regular API calls, referencing the compiled structures.  It can also be used on a per-call basis for a moderate speed up without application changes.

In this ticket, we only allow compilation of configuration used in WT_SESSION::begin_transaction.

This change also includes various configuration parsing performance improvements.

ddanderson commented:
evergreen retry

ddanderson commented:
@korteland, I've invited you to be an initial reviewer.  There's some rough edges, but pretty much everything is there.  After a review pass or two, I'll invite more to look at this.  You could start at `test/csuite/wt11126_compile_config` to see how the application does the precompiling.  And you could look at `__wt_txn_config` to see how WT "mainline" code works with the new conf system.  `dist/api_config_gen.py` builds up the tables used for preconfiguration - but probably better to look at its output: changes to `config_def.c` and the new file `conf.h` (look at the data structures in `conf.h` and new changes to `config.h`).  With that as a foundation, looking at `src/conf/*.c` will make more sense.  Thanks!

ddanderson commented:
Also good to realize that the application _can_ do precompiling, but even if it doesn't, upon entry to the begin_transaction API, the regular configuration string will be compiled then, and the compiled structures will be used for the duration of the call.  For this test case, even without the caller precompiling, we get about a 50% speedup.  With precompiling, its about 7x.  The configuration that MDB uses for begin_transaction uses 4 keys, and 2 of them are in a nested configuration.  The length and relative complexity of the config is the reason that the old config system is relatively slow and we can get an impressive speedup.

ddanderson commented:
evergreen retry

korteland commented on src/config/config_def.c:
Apologies for slacking off, but I'm not particularly willing to review this file ;-)

korteland commented on src/txn/txn.c:
```diff
@@ -536,12 +536,46 @@ __txn_config_operation_timeout(WT_SESSION_IMPL *session, const char *cfg[], bool
+ *     Configure a transactions operation timeout duration.
+ */
+static int
+__txn_conf_operation_timeout(WT_SESSION_IMPL *session, const WT_CONF *conf, bool start_timer)
```
There's nothing wrong with this particular function, but I'm more asking in general - is there going to be a pattern of functions accepting both existing `char *cfg` and new `WT_CONF *conf` arguments? If so, it's probably worth looking into some way to avoid repeating too much code.

ddanderson replied:
Having two functions, one for config, one for conf, is generally the exception.  The issue here is that the original function `__txn_conf_operation_timeout` was used by: begin_transaction, commit_transaction, rollback_transaction.  If/when we convert begin_transaction and commit_transaction to the new way, we can get rid of the extra function. 

Sorry I didn't reply to this sooner!

korteland commented on src/include/conf.h:
```diff
@@ -0,0 +1,1190 @@
+    struct {
+        WT_CONF_BIND_DESC *desc;
+        WT_CONFIG_ITEM item;
+    } values[WT_CONF_BIND_VALUES_LEN];
```
Does this restrict the number of bound values to five per API call?

korteland commented on src/include/conf.h:
```diff
@@ -0,0 +1,1190 @@
+ */
+struct __wt_conf {
+    /*
+     * Explicit compilation is allocated on the heap, implicit is allocated on the stack.
```
Can I ask you for a short explainer (or a pointer to a comment that I might've missed) about the position-independence thing and and the way it interacts with explicit/implicit items and heap/stack allocation? I'm finding it hard to piece together from the code.

korteland commented on src/txn/txn.c:
```diff
@@ -536,12 +536,46 @@ __txn_config_operation_timeout(WT_SESSION_IMPL *session, const char *cfg[], bool
+ *     Configure a transactions operation timeout duration.
+ */
+static int
+__txn_conf_operation_timeout(WT_SESSION_IMPL *session, const WT_CONF *conf, bool start_timer)
```
Another thread that I'll hang off this new function because it's a convenient spot - the convention seems to be "old style -> `cfg`/`config`, new style -> `conf`", at least as a rule of thumb. I know discussing naming on a PR is a rabbit hole, but I suspect I'm going to find it tricky to keep this in my head and differentiate them consistently. Have you considered a name that more clearly says "precompiled", e.g. `pconf`?

ddanderson replied:
I'm okay with any alternate naming.

korteland commented on src/include/conf.h:
Another bigger, design-y question: I can see we've now got a big list of structs that we fill in behind the scenes that contains a strongly-typed version of the parsed/compiled string config. Have you considered exposing these structures as part of the API so someone could do e.g.

```
opts = struct session_opts {
    .cache_cursors = false,
    .ignore_cache_size = true
};
conn->open_session(conn, NULL, NULL, &opts);
```

(I'm not suggesting you rewrite it like this, just trying to get at the motivations for picking the compilation approach over other methods.)

ddanderson replied:
Thanks for that insight.  Hmm, honestly I didn't think deeply about this approach.  It certainly is readable.  I think one motivation to continue using a string - was to not change the type of the argument for the various API functions that might eventually use this.  To get full type checking on the argument, we'd need new functions if we wanted to support the old function.  For example, `conn->open_session_opts()` for this new one.  I honestly didn't know much about that C99 syntax.  The fact that it creates a zeroed struct otherwise is huge, I didn't know that until I just looked it up.  There are some oddities about our current arrangement that we'd need to deal with.  For example `.cache_cursors` is by default `true`.  So the example above is the same as:
```
opts = struct session_opts {
    .ignore_cache_size = true
};
```
We need to distinguish an explicit assignment to false (or 0) with the implicit zeroing of the struct. If we renamed that (and other fields like it), e.g. `.no_cached_cursors` would fix it.  Config strings have a lot of flexibility (for good and evil!), so there are probably some issues to sort out.  There's probably lots of little edge cases that require various workarounds.  But I bet it's doable.

With the "struct" approach, you create another code path in the API.  Internally, you'd probably want to rewrite `__conn_open_session` to call the new `__conn_open_session_opt` to minimize redundant code - which makes a lot of sense.

Precompiling does have some complexity, but once we have it committed, you have an easier path (I think) to convert a new API to use it - that is, there's less WT code that is affected.  I think I was drawn to that.

korteland reviewed (commented):
This is really impressive, and the performance gains even more so. I've got a few higher-level/design questions before drilling into the code itself if that's OK?

ddanderson commented on cmake/strict/strict_flags_helpers.cmake:
```diff
@@ -65,7 +65,7 @@ function(get_gnu_base_flags flags)
         list(APPEND gnu_flags "-Wduplicated-cond")
         list(APPEND gnu_flags "-Wlogical-op")
-        list(APPEND gnu_flags "-Wunused-const-variable=2")
+        list(APPEND gnu_flags "-Wunused-const-variable=1")
```
This change is needed to allow the "trick" used by `WT_CONF_ID_STRUCTURE` in conf.h.  Set as `1` allows `static const` variables that are not used to be allowed in include files, but you get a warning/error if it in the compilation unit itself.  Set as `2` gives warnings for both.  I think `1` is reasonable, as it would allow us to replace some #define constants with typed static constants.

This change could be split out if it warrants more discussion.

ddanderson commented on src/conf/conf_bind.c:
```diff
@@ -0,0 +1,79 @@
+
+            /*
+             * Even when the bind format uses %s, we must check it against the numeric and boolean
+             * values, as the base config parser does the same.
```
The base config parser doesn't have much context.  It sees "x=true", and it doesn't know if x will be interpreted as a boolean variable or a string.  So it returns a config item with str set to "true", len = 4 and val = 1.  The WT core code can treat it as a string or boolean and it works either way.

ddanderson replied:
@pmacko86 I'm hoping that the arch guide page that I made would give some general coverage of what's going on here.  Do you think we should a small or large summary in conf.h?  And/or a pointer to the arch guide?  (We've never done that before - doesn't mean we shouldn't).

ershov commented on test/csuite/wt11126_compile_config/main.c:
```diff
@@ -0,0 +1,329 @@
+/*-
```
A few suggestions for safety testing:
* From the existing Python metadata generate a test that would set every config key, then check that the values that you get from the compiled version are the same.
* Check every key's default value that you get if it's unset.
* Check incorrect inputs for various types (wrong values in place of integers or booleans or subconfigs).
* What happens if the user passes a string formatted as a compiled config? Since there's no distinction of the config type, we handle the compiled variant the same way as clear text one which can be unsafe.

pmacko86 replied:
Agree that the test should verify that the implementation is doing the right thing with error checking. I'm not sure we necessarily need to check every single key, as they are automatically generated, but there should be enough coverage to ensure that we are doing the right thing.

marcbutler commented on src/include/misc.h:
```diff
@@ -312,6 +312,14 @@ __wt_string_match(const char *str, const char *bytes, size_t len)
+ * This is a faster version of WT_STRING_MATCH, and should be used where performance matters. bytes
+ * or str should never be NULL.
+ */
+#define WT_FAST_STRING_MATCH(str, bytes, len)                                                   \
```
In the case `len==0` even assuming `str` and `bytes` are both valid pointers to readable memory; this will make the invalid comparison `str[0]==bytes[0]` potentially resulting in an evaluation of false.

```suggestion
#define WT_FAST_STRING_MATCH(str, bytes, len)                                                   \
 ((len==0) || (((const char *)(str))[0] == ((const char *)(bytes))[0] && strncmp(str, bytes, len) == 0 && \
      (str)[len] == '\0'))
```

ddanderson replied:
@marcbutler - In the limited uses of this macro, I think at least the first string is non-zero length, so it shouldn't trigger the invalid case.  I'll add an assert and comment.

ddanderson replied:
In the cases I'm interested in, the first arg is known to have non-zero length.  I made a WT_RESTRICTED_STRING_MATCH that uses an assert to check the constraints.  Let me know what you think.

pmacko86 commented on test/csuite/wt11126_compile_config/main.c:
```diff
@@ -0,0 +1,329 @@
+
+    printf("number of calls: %d\n", N_CALLS * N_RUNS);
+    base_ns = ns = nsecs[0] / (N_CALLS * N_RUNS);
+    for (variant = 0; variant < N_VARIANTS; ++variant) {
```
Nice that we are collecting numbers on this!

pmacko86 commented on src/conf/conf_compile.c:
```diff
@@ -0,0 +1,517 @@
+ *     Compare function used for binary search.
+ */
+static int
+__conf_check_compare(const void *keyvoid, const void *check)
```
Having two different types for left and right keys is not safe, as it makes assumptions about the implementation of `bsearch`. (As a point in case, implementations of the equivalent C++ functions do not assume that the right parameter is the "needle" value that you are searching for.)

pmacko86 replied:
According to the description in https://en.cppreference.com/w/c/algorithm/bsearch, it seems that this behavior is allowed. Hopefully this is indeed a part of the standard! :)

ddanderson replied:
Yes, I think this is part of the standard.  I see in
https://pubs.opengroup.org/onlinepubs/9699919799/
```
The comparison function pointed to by compar shall be called with two arguments that point to the key object and to an array element, in that order.
```
I'll add to the comment.

pmacko86 reviewed (commented):
Nice piece of work. I have mostly small comments; the only two bigger comments are that we should have more testing around error handling, and that the binary search during compilation is not safe.

I admit I didn't quite follow everything, but I suppose that's why we have several reviewers :)

ddanderson commented:
@pmacko86 @ershov @marcbutler, thank you all for comments thus far.  I've put together an arch doc page (now part of this PR) and will attach a pdf for it here.  Peter and Yury, this is updated from what you've seen, now I think the arch doc can be considered complete. I hope it will help.
[ConfigurationStringCompilation3.pdf](https://github.com/wiredtiger/wiredtiger/files/14014807/ConfigurationStringCompilation3.pdf)

ddanderson commented on dist/api_config.py:
```diff
@@ -179,7 +221,74 @@ def getconfcheck(c):
+        add_keys(keynumber, config, '')
+
 if not test_config:
+    for name in sorted(config_names):
```
Done

ddanderson commented on dist/api_config.py:
```diff
@@ -179,7 +221,74 @@ def getconfcheck(c):
+        # For names that do not have a class (e.g. wiredtiger_open), place it in GLOBAL
+        return ('GLOBAL', name)
+
+for name in api_data_def.methods.keys():
```
Nice, I fixed a few places with the same construct.

ddanderson commented on dist/api_config.py:
```diff
@@ -28,6 +28,37 @@
+        if not name in self.numbering:
+            num = self.count
+            self.count += 1
+            self.numbering[name] = num
```
Done.

ddanderson commented on dist/api_config.py:
```diff
@@ -498,9 +619,166 @@ def getsubconfigstr(c):
+
+    with conf_h.replace_fragment('Per-API configuration structure declarations') as tfile:
+        method_to_counts = dict()
+        for name in api_data_def.methods.keys():
```
Fixed

ddanderson commented on dist/api_config.py:
```diff
@@ -498,9 +619,166 @@ def getsubconfigstr(c):
+    # Assign unique numbers to all keys used in configuration, regardless of "level".
+    with conf_h.replace_fragment('API configuration keys') as tfile:
+        count = 0
+        a = keynumber.numbering
```
Fixed

ddanderson commented on dist/api_config.py:
```diff
@@ -498,9 +619,166 @@ def getsubconfigstr(c):
+    with conf_h.replace_fragment('API configuration keys') as tfile:
+        count = 0
+        a = keynumber.numbering
+        b = dict()
```
Fixed

ddanderson commented on dist/api_config.py:
```diff
@@ -498,9 +619,166 @@ def getsubconfigstr(c):
+            (nconf, nitem) = method_to_counts[name]
+            if nitem == 0:
+                continue
+            if '.' in name:
```
Yes!

ddanderson commented on src/config/config_api.c:
```diff
@@ -320,10 +320,6 @@ __wt_configure_method(WT_SESSION_IMPL *session, const char *method, const char *
         WT_RET_MSG(session, EINVAL, "no configuration specified");
     if (type == NULL)
         WT_RET_MSG(session, EINVAL, "no configuration type specified");
-    if (strcmp(type, "boolean") != 0 && strcmp(type, "int") != 0 && strcmp(type, "list") != 0 &&
```
@pmacko86 There is a check further down in this function that covers the same thing.  I added the later code in WT-11440, but forgot to remove the redundant check then.

ddanderson commented on src/conf/conf_bind.c:
```diff
@@ -0,0 +1,86 @@
+            value->str = str;
+            value->len = len;
+
+            /*
+             * Even when the bind format uses %s, we must check it against boolean constants. This
+             * is done for the non-compiled cases and WiredTiger configuration processing code may
+             * depend on it. We also change the string to point to fixed values of these constants
+             * to have a consistent way to fast match strings that are part of choices. "false" and
+             * "true" are legal as parts of a set of choices, and so can be used with
+             * WT_CONF_STRING_MATCH.
+             */
```
@pmacko86, does this comment make more sense?

pmacko86 replied:
Yes, this helps, thank you. I'd also add something like "The resulting configuration item can be subsequently interpreted both as a boolean or as a string."

What was confusing for me is that if someone does something strange like passing the equivalent of `"name=true"` to `WT_SESSION::checkpoint`, it wouldn't fail because the name is a boolean and indeed create a checkpoint named "true", because the `name` can be interpreted as both a string and a boolean (depending on what the implementation needs). This is basically what I thought would be an important point to clarify.

ddanderson commented:
@ershov and @pmacko86, can you please review again?

I've addressed the review comments and extended the testing by adding a test/csuite/config test.  The difficulty in exhaustive testing is how do you know if a configuration has the desired effect?  Some config flags we can see by peering into the WT_TXN struct to see flag values (that's what test/csuite/wt11126_compile_config does).  But some will probably be revealed by correct behavior in the Python test suite, which exercises various begin_transaction configurations (and now we precompile all configs for begin_transaction).

To be more certain, I added verbose messages when compiling, showing the inputs and then creating a configuration "output" string from the compiled_configuration.  test/csuite/config uses that to poke at every config key and make sure that valid and invalid inputs do the right thing.  It does some additional miscellaneous tests as well.

pmacko86 commented on test/csuite/config/main.c:
```diff
@@ -0,0 +1,494 @@
+    free(state->method_name);
+    state->method_name = NULL;
+    for (i = 0; i < MAX_INPUT_CONFIG; ++i) {
+        free(state->input_config[i]);
```
Are we sure that input config is always not NULL?

ddanderson replied:
`input_config` entries may be NULL, but `free(NULL)` must be accepted by a conforming C compiler.

ddanderson commented:
@pmacko86 and @ershov, I just made some changes for readability (I hope!).
- Almost all the generated boilerplate has been moved from conf.h to conf_keys.h .  conf.h now contains struct definitions and a little bit of generated stuff that must appear after the struct definitions.
- There used to be a `WT_CONF_KEY` struct.  You get to it from looking up via a key, but it actually _contains_ the value that will be returned.  So I renamed it to be `WT_CONF_VALUE`.  And what used to be the `key_map` to get to the right one, is now the `value_map`.  I made corresponding fixes to the arch guide.

@pmacko86, this doesn't need a re-review (unless you want), these all mechanical changes.  Thanks for your review, BTW!

ddanderson commented on dist/api_config.py:
```diff
@@ -498,9 +616,157 @@ def getsubconfigstr(c):
+        for name in sorted(api_data_def.methods.keys()):
+            # Extract class and method name
+            (clname, mname) = get_method_name_parts(name)
+            count += 1
```
The code is misleading (even if correct), it should print the count of number of elements in the array.  So I changed it to do that directly, rather than have a counter.  I added a comment too.  Thanks for spotting that.

ddanderson commented on src/conf/conf_get.c:
```diff
@@ -0,0 +1,91 @@
+        partkey = keys & 0xffff;
+        WT_ASSERT(session, partkey != 0);
+
+        conf_key_index = conf->key_map[partkey];
```
Yes, I added an assertion here.

ddanderson commented on src/conf/conf_compile.c:
```diff
@@ -0,0 +1,711 @@
+            if (check_type != WT_CONFIG_ITEM_NUM && check_type != WT_CONFIG_ITEM_BOOL)
+                WT_RET_MSG(session, EINVAL, "Value '%.*s' is not compatible with %s type",
+                  (int)value->len, value->str, check->type);
+        } else if (value->str[1] == 's') {
```
I added a check of len >= 2, returns an error w/ message (it implies caller passed us something like `"foo=%"`)

ddanderson commented on src/conf/conf_compile.c:
```diff
@@ -0,0 +1,711 @@
+    conn->conf_array[compiled_entry] = conf;
+
+    /*
+     * Mark the entry so it won't be mistakenly freed when used for API calls.
```
That was an obsolete comment.  I replaced it with a comment that describes what we are returning.  Thanks!

ddanderson commented on src/conf/conf_compile.c:
```diff
@@ -0,0 +1,711 @@
+    /*
+     * Mark the entry so it won't be mistakenly freed when used for API calls.
+     */
+
```
Fixed.

ddanderson commented on src/conf/conf_compile.c:
```diff
@@ -0,0 +1,711 @@
+    WT_RET(__wt_calloc_def(session, conn->conf_max, &conn->conf_array));
+    conn->conf_size = 0;
+
+    WT_RET(__wt_calloc(session, WT_CONF_API_ELEMENTS, sizeof(WT_CONF *), &conn->conf_api_array));
```
Yes it does.  I added a comment to make it clear.

ershov commented on test/csuite/config/main.c:
```diff
@@ -0,0 +1,494 @@
+/*-
```
Can we have tests for overlapping keys in sub-configs and various types of incorrect inputs?

Also, it'd be great to add a test calling an API with an "unbound config".

ddanderson replied:
I added a test for unbound config (in test/csuite/config/main.c).

I'm not sure what you mean by "overlapping keys in sub-configs".  Eventually, when we support more APIs, and there are keys known by multiple APIs, we'll want to check that.  But right now, begin_transaction just has a single sub-config ("roundup_timestamps") with two possible keys ("prepared", "read"), and test/csuite/config tests valid and invalid values for each.

test/csuite/config/main.c tests lots of incorrect inputs - check_compiling_invalid_configurations() sets every key to multiple invalid values.  Is there something else you're thinking of?

ddanderson replied:
Re: `"cconf"` for compiled config.  There are 20 places that `WT_CONF *conf` declaration appears, and each has several uses.  I'm okay with changing them if you think it helps readability - just wanted to make sure that's what you're looking for.

ershov replied:
> Re: `"cconf"` for compiled config. There are 20 places that `WT_CONF *conf` declaration appears, and each has several uses. I'm okay with changing them if you think it helps readability - just wanted to make sure that's what you're looking for.

Thanks Don.
I was also thinking about function names. Plain "conf" looks to me like a contraction of the word "config" and it's not apparent how it's different. Something like "cconf" adds that special bit to it.

ddanderson replied:
Thanks, I'm going to push this suggestion to a followup ticket, in the interest of getting the baseline of work done.

ershov reviewed (commented):
Good job Don!

ddanderson commented:
@ershov, any chance we can get this through review this week?  I'd like to get it committed at least a week before I'm on vacation.

ershov reviewed (approved):
Hey Don, thank you for this big work.
I'm ok with this PR, although one corner case test that I would like to see is - compiling and accessing subconfig keys with the same name like "sync" that exists in multiple subconfigs.

ddanderson commented:
> Hey Don, thank you for this big work. I'm ok with this PR, although one corner case test that I would like to see is - compiling and accessing subconfig keys with the same name like "sync" that exists in multiple subconfigs.

@ershov Thanks for the review!  We cannot compile any string other than ones that are used by begin_transaction (the test/csuite/config test confirms this by trying to compile a string for WT_SESSION->create at main.c:430).
In that test, we test every possible key for begin_transaction (including `sync`), most are table driven in the test.  I locally added a printf at main.c:344, and here are the calls to compile_configuration that are table driven:
```
Compiling expect failure: ignore_prepare=foo
Compiling expect failure: isolation=foo
Compiling expect failure: name=+$
Compiling expect failure: no_timestamp=foo
Compiling expect failure: no_timestamp=TRUE
Compiling expect failure: no_timestamp=FALSE
Compiling expect failure: no_timestamp=-2
Compiling expect failure: operation_timeout_ms=foo
Compiling expect failure: operation_timeout_ms=
Compiling expect failure: operation_timeout_ms=-42
Compiling expect failure: priority=foo
Compiling expect failure: priority=-101
Compiling expect failure: priority=101
Compiling expect failure: read_timestamp=+$
Compiling expect failure: roundup_timestamps=(prepared=foo)
Compiling expect failure: roundup_timestamps=(prepared=TRUE)
Compiling expect failure: roundup_timestamps=(prepared=FALSE)
Compiling expect failure: roundup_timestamps=(prepared=-2)
Compiling expect failure: roundup_timestamps=(read=foo)
Compiling expect failure: roundup_timestamps=(read=TRUE)
Compiling expect failure: roundup_timestamps=(read=FALSE)
Compiling expect failure: roundup_timestamps=(read=-2)
Compiling expect failure: sync=foo
Compiling expect failure: sync=TRUE
Compiling expect failure: sync=FALSE
Compiling expect failure: sync=-2
Compiling expect success: ignore_prepare=true
Compiling expect success: ignore_prepare=false
Compiling expect success: ignore_prepare=force
Compiling expect success: ignore_prepare=
Compiling expect success: isolation=read-uncommitted
Compiling expect success: isolation=read-committed
Compiling expect success: isolation=snapshot
Compiling expect success: isolation=
Compiling expect success: name=foo
Compiling expect success: name=
Compiling expect success: name=42
Compiling expect success: name=-42
Compiling expect success: no_timestamp=true
Compiling expect success: no_timestamp=false
Compiling expect success: no_timestamp=0
Compiling expect success: no_timestamp=1
Compiling expect success: no_timestamp=
Compiling expect success: operation_timeout_ms=42
Compiling expect success: operation_timeout_ms=0
Compiling expect success: priority=0
Compiling expect success: priority=-100
Compiling expect success: priority=100
Compiling expect success: read_timestamp=foo
Compiling expect success: read_timestamp=
Compiling expect success: read_timestamp=42
Compiling expect success: read_timestamp=-42
Compiling expect success: roundup_timestamps=(prepared=true)
Compiling expect success: roundup_timestamps=(prepared=false)
Compiling expect success: roundup_timestamps=(prepared=0)
Compiling expect success: roundup_timestamps=(prepared=1)
Compiling expect success: roundup_timestamps=(prepared=)
Compiling expect success: roundup_timestamps=(read=true)
Compiling expect success: roundup_timestamps=(read=false)
Compiling expect success: roundup_timestamps=(read=0)
Compiling expect success: roundup_timestamps=(read=1)
Compiling expect success: roundup_timestamps=(read=)
Compiling expect success: sync=true
Compiling expect success: sync=false
Compiling expect success: sync=0
Compiling expect success: sync=1
Compiling expect success: sync=
```
